library/bignum_core.c

```diff
@@ -33,8 +33,28 @@
 #include "bn_mul.h"
 #include "constant_time_internal.h"
 
-size_t mbedtls_mpi_core_clz(mbedtls_mpi_uint a)
+inline size_t mbedtls_mpi_core_clz(mbedtls_mpi_uint a)
 {
+    /* Note: the result is undefined for a == 0
+     * (because this is the behaviour of __builtin_clz).
+     */
+#if defined(__has_builtin)
+#if __has_builtin(__builtin_clz)
+    if (sizeof(mbedtls_mpi_uint) == sizeof(unsigned int)) {
```

Comment:
That is a good idea, it seems clang and gcc support `__builtin_clzll` as far back as I can test (gcc 4.5.4 and clang 3.0) so I've simplified this (and also added MSVC support).

Reply:
I don't like the first suggestion because it essentially relies on accurately recreating the logic from `bignum.h` to select the correct option, which is brittle (e.g., would break if we decided this type should always be 64 bits, or if we added an option for the user to specify the size, etc). Comparing via sizeof is much more robust (and still a compile-time decision).

Reply:
I guess we could define `MBEDTLS_MPI_BITS` to 32 or 64 in `bignum.h`. Then, here we could do:

```
#if defined(MBEDTLS_CLZ32) && (MBEDTLS_MPI_BITS == 32)
    return MBEDTLS_CLZ32(a);
#elif defined(MBEDTLS_CLZ64) && (MBEDTLS_MPI_BITS == 64)
    return MBEDTLS_CLZ64(a);
#else
    // plain C impl
#endif
```

with some suitable #defines for MBEDTLS_CLZ32 outside of the function. Maybe that's a bit cleaner? I don't have a strong perference.

Reply:
> I don't like the first suggestion because it essentially relies on accurately recreating the logic from bignum.h to select the correct option

I don't understand: what logic would we be recreating? `bignum.h` guarantees that `MBEDTLS_HAVE_INT32` is defined if `mbedtls_mpi_uint` is a 32-bit type, and similar for `MBEDTLS_HAVE_INT64`. If we added support for a different size, then neither symbol would be defined and instead we'd define e.g. `MBEDTLS_HAVE_INT16`. `defined(MBEDTLS_HAVE_INTxx)` vs `MBEDTLS_MPI_BITS == xx` is just a cosmetic difference since we aren't doing anything with `xx` other than an equality comparison.

Reply:
> bignum.h guarantees that MBEDTLS_HAVE_INT32 is defined (…)

Mind you, this is not clearly documented (and somewhat surprising given the name of the symbols) — I only realized that while doing bignum recently. So it isn't a promise we make to our users as part of our API stability guarantee. But it's a property that we are already relying on, and there is a precedent (I forget where) of relying on this property to simplify some bignum code.

Reply:
> this is not clearly documented (and somewhat surprising given the name of the symbols)

Yes, that was not obvious to me. In that case I'm happy to switch on that instead of `MBEDTLS_MPI_BITS`, but no strong preference vs. the current arrangement.

Reply:
Note: I agree with Gilles here: `MBEDTLS_HAVE_INTxx` means precisely that `mbedtls_mpi_uint` is xx bits. Yes, the naming is a bit off

Reply:
However, while we'd normally use `MBEDTLS_HAVE_INTxx`, using `sizeof()` as done here is perfectly natural and leads to a quite readable implementation, so I think this is fine. If we need 32- and/or 64-bit `clz()` functions, we can create those then then